source/common/upstream/load_balancer_impl.cc

```diff
@@ -437,20 +432,40 @@ void RoundRobinLoadBalancer::refresh(uint32_t priority) {
       }
     }
 
+    // Compute schedule offset for unweighted before deleting the existing
+    // scheduler.
+    uint64_t unweighted_offset = 0;
+    if (hosts.size() > 0) {
+      // If we already have been balancing for this locality, continue where we
+      // left off; a rebuild with the same hosts will have the expected RR
+      // across the rebuild. Otherwise, start with the LB seed.
+      if (scheduler_.find(source) != scheduler_.end()) {
+        unweighted_offset = scheduler_[source].rr_index_;
+      } else {
+        unweighted_offset = seed_ % hosts.size();
+      }
+    }
     // Nuke existing scheduler if it exists.
-    scheduler_[source] = Scheduler{};
-    scheduler_[source].weighted_ = weighted;
+    auto& scheduler = scheduler_[source] = Scheduler{};
+    scheduler.weighted_ = weighted;
     if (weighted) {
       // Populate scheduler with host list.
-      // TODO(htuch): We should add the ability to randomly offset into the host list to
-      // desynchronize the schedule across Envoys in large fleets.
       for (const auto& host : hosts) {
         // We use a fixed weight here. While the weight may change without
         // notification, this will only be stale until this host is next picked,
         // at which point it is reinserted into the EdfScheduler with its new
         // weight in chooseHost().
-        scheduler_[source].edf_.add(host->weight(), host);
+        scheduler.edf_.add(host->weight(), host);
       }
+      // Cycle through hosts to achieve the intended offset behavior.
+      // TODO(htuch): Consider how we can avoid biasing towards earlier hosts in the
+      // schedule across refreshes for the weighted case.
+      for (uint32_t i = 0; i < seed_ % hosts.size(); ++i) {
```

Comment:
A lot of this behavior in the EDF case is not intuitive to me at all. Not a big deal right now but is it worth at some point writing some LB simulations to cover this case so we can better understand the behavior? I wonder for example if we really have to cycle through all hosts to get reasonable randomness since doing that might be costly for large host sets?

Reply:
I agree, EDF is pretty confusing and I think we might want to move away from it longer term. I've filed https://github.com/envoyproxy/envoy/issues/3278 and will merge as is.